### ISSUE TITLE
consider calling syscall.Exec

### ISSUE THREAD

**Issue by kevinburke** [segmentio/chamber#37]
execRun does a ton of work to pass through stdin, stdout, signals, return codes, etc from a single child process.

Instead of doing that I am wondering if it would be easier to simply call `syscall.Exec` with the arguments you want. That would replace the current process with the child process, and you wouldn't need to worry about signals or inputs/outputs being passed through correctly.

https://golang.org/pkg/syscall/#Exec

**Comment by achille-roussel** [comment]
Seems like a good idea to me, the main motivation would be situations where the process gets SIGKILL, if chamber sits in between it would get killed, then its child process would get attached to init... doesn't seem ideal.
I think the only downside would be that `syscall.Exec` isn't available when compiling for Windows, we would probably want to keep building with the current implementation in this case.

**Comment by kevinburke** [comment]
Alright, now I'm confused. Ian Lance Taylor says it's unnecessary to forward signals, https://go-review.googlesource.com/c/go/+/72550

However both this lib and aws-vault try to proxy signals sent from a parent to the child. Would that result in the signals being delivered twice?

**Comment by dfuentes** [comment]
Pretty easy to test.  https://gist.github.com/dfuentes/c326e8f496e9e16c2780b911880c06fe contains two stripped down files, one spins up a subprocess and waits on it (forwarding signals if the FORWARD env var is set), the other waits to receive a signal and prints it.

In my environment, running without forwarding works for SIGINT, but doesn't work for SIGTERM.  When I SIGTERM the parent process without forwarding, the parent exits and the orphan process is reparented and continues running.

I'd vote for just leaving the code as is, since we know it works, and I don't see a compelling reason to change it.

**Comment by dfuentes** [comment]
Chamber does forward it.  I was saying the PoC code I linked (when set to not forward signals) does not.

**Comment by achille-roussel** [comment]
Oh OK I misunderstood.

Another compelling reason would be proper handling of SIGSTOP, which also cannot be caught by processes. Sending SIGSTOP to a container running a chamber process pauses chamber instead of pausing the child process, which is very likely not the expected behavior. It could also have some concerning side effects since having chamber paused but the child process running, `os/exec` uses goroutines to pipe standard inputs and outputs between the parent and child processes, if the parent is paused then the child process won't be able to produce logs for example, even tho it's actually running.

Another limitation of the current implementation is chamber only forwards SIGINT and SIGTERM (note that `os.Kill` is SIGKILL which cannot be intercepted). Which means processes that use other signals will not be able to receive them if they are sent to chamber instead of directly to the process (think using `docker kill -s SIGUSR1` for example, which we use for temporarily enabling debug logs https://github.com/segmentio/events/blob/master/sigevents/init.go).

It seems like chamber would be much more robust if it was not getting in the way of the program it's setting up the environment for, and just hand off the execution once it's done its work.

**Comment by dfuentes** [comment]
Seems like we could/should add SIGUSR and friends to the notify list.  SIGSTOP seems like an edge case we are unlikely to see (docker itself doesn't use SIGSTOP, it uses the cgroups freeze system instead, which should work on the full hierarchy of processes).  The advice from Rob Pike in the above linked conversation is also not to use exec https://go-review.googlesource.com/c/go/+/72550#message-2a929c7c209e61370bab84406b1569f8ccc757ca.

That said, I'm more than happy to review a PR to use exec instead.

**Comment by kevinburke** [comment]
I want to put a proposal to the Go team add a new/better API here. I don't think they are quite aware of the problem. I also wrote a little more on this here https://kev.inburke.com/kevin/proxying-to-a-subcommand-with-go/

**Comment by achille-roussel** [comment]
My point wasn't so much to whitelist SIGUSR* but more that we can't really plan what signal a process is going to rely on.

I think this is worth solving, I'll send a PR.

**Comment by lox** [comment]
> Alright, now I'm confused. Ian Lance Taylor says it's unnecessary to forward signals, https://go-review.googlesource.com/c/go/+/72550 

Sounds like he's talking about process groups. He's correct, subprocesses inherit the same pgid and then if a signal is sent to the process group then everything with the pgid will get it. Signals sent directly to the parent process via pid won't be propagated to the subprocess as far as I know.

WRT to syscall.Exec, I tried that several times with `aws-vault` and ended up going with the proxying approach for portability and various edge cases we ran into.
